### PR TITLE
Form Helper maxlength attribute uses validation #12773

### PR DESCRIPTION
**Pull request by lamasgergo**
Improvement for 
https://github.com/cakephp/cakephp/issues/12773

Form helper now supports maxlength attribute from validation directly.

### CODE REVIEW

File: src/View/Form/ArrayContext.php [hunk 1]
```diff
@@ -206,6 +206,27 @@ public function isRequired($field)
         return (bool)$required;
     }
 
+    /**
+     * Get field length from validation
+     *
+     * In this context class, this is simply defined by the 'maxlength' array.
+     *
+     * @param string $field A dot separated path to check required-ness for.
+     * @return int|false
```
Comment (dereuromark) [inline src/View/Form/ArrayContext.php]: for new API methods i would rather want to see nullable 

    int|null

also for future php typehinting

File: src/View/Form/ContextInterface.php [hunk 2]
```diff
@@ -70,6 +70,14 @@ public function val($field);
      */
     public function isRequired($field);
 
+    /**
+     * Get maximum length of a filed from model validation
+     *
+     * @param string $field A dot separated path to check required-ness for.
+     * @return bool|false
```
Comment (dereuromark) [inline src/View/Form/ContextInterface.php]: same

File: src/View/Form/ArrayContext.php [hunk 3]
```diff
@@ -212,19 +213,20 @@ public function isRequired($field)
...
         }
         $fieldShema = Hash::get($this->_context['schema'], $field);
-        if(is_array($fieldShema)) {
+
+        if (is_array($fieldShema)) {
             $maxLength = Hash::get($fieldShema, 'length');
         }
 
-        return $maxLength ?? false;
+        return $maxLength ?? null;
```
Comment (dereuromark) [inline src/View/Form/ArrayContext.php]: isnt $maxLength already nullable? The cloaking ?? should be avoided here then.

File: src/View/Form/EntityContext.php [hunk 4]
```diff
@@ -432,16 +432,17 @@ public function isRequired($field)
      * @param string $field The dot separated path to the field you want to check.
      * @return int|false
```
Comment (dereuromark) [inline src/View/Form/EntityContext.php]: null?

**Comment by dereuromark** [comment]
@lamasgergo This cannot go into 3.6, this should go into 3.7 (next).
But we cannot fix the existing methods to nullable I think - this might be too much a BC break.
So better to stick with "false" for now as return value, sry.
We can fix this hopefully for 4.0 then.

File: src/View/Form/ArrayContext.php (hunk 3 above)
Comment (dereuromark) [reply src/View/Form/ArrayContext.php]: Actually, since those are existing API methods, better to keep false, see below.

**Comment by lamasgergo** [comment]
> @lamasgergo This cannot go into 3.6, this should go into 3.7 (next).
> But we cannot fix the existing methods to nullable I think - this might be too much a BC break.
> So better to stick with "false" for now as return value, sry.
> We can fix this hopefully for 4.0 then.

So revert to false returns everywhere?

File: src/View/Form/ArrayContext.php [hunk 5]
```diff
@@ -206,6 +207,28 @@ public function isRequired($field)
...
+     */
+    public function getMaxLength($field)
+    {
+        if (!is_array($this->_context['schema'])) {
+            return null;
+        }
+        $fieldShema = Hash::get($this->_context['schema'], $field);
+
+        if (is_array($fieldShema)) {
+            $maxLength = Hash::get($fieldShema, 'length');
```
Comment (dereuromark) [inline src/View/Form/ArrayContext.php]: $maxLength is not defined before, but it should be.

**Comment by dereuromark** [comment]
Please, after setting the target to 3.next, add some test cases for this new behavior.

**Comment by dereuromark** [comment]
> So revert to false returns everywhere?

No, I just didnt see the correct diff :) My mistake. New API methods, all good.

**Comment by lamasgergo** [comment]
> Please, after setting the target to 3.next, add some test cases for this new behavior.

I have added 3 test case into FormHelperTest. Isnt it this enough?

File: src/View/Form/ContextInterface.php [hunk 6]
```diff
@@ -70,6 +70,14 @@ public function val($field);
      */
     public function isRequired($field);
 
+    /**
+     * Get maximum length of a field from model validation
+     *
+     * @param string $field A dot separated path to check required-ness for.
+     * @return bool|null
```
Comment (dereuromark) [inline src/View/Form/ContextInterface.php]: why bool?

File: src/View/Form/EntityContext.php [hunk 7]
```diff
@@ -426,6 +426,27 @@ public function isRequired($field)
...
+    public function getMaxLength($field)
+    {
+        $parts = explode('.', $field);
+        $validator = $this->_getValidator($parts);
+        $fieldName = array_pop($parts);
+        if (!$validator->hasField($fieldName)) {
+            return null;
+        }
+        foreach ($validator->field($fieldName)->rules() as $rule) {
+            if ($rule->get('rule') == 'maxLength') {
```
Comment (dereuromark) [inline src/View/Form/EntityContext.php]: `===` always for strings

File: src/View/Form/EntityContext.php [hunk 8]
```diff
@@ -426,6 +426,27 @@ public function isRequired($field)
...
+                return $rule->get('pass')[0] ?? null;
+            }
+        }
```
Comment (dereuromark) [inline src/View/Form/EntityContext.php]: missing return null
semantically, this would be void otherwise.

File: src/View/Form/FormContext.php [hunk 9]
```diff
@@ -134,6 +134,24 @@ public function isRequired($field)
...
+     * {@inheritDoc}
+     */
+    public function getMaxLength($field)
+    {
+        $validator = $this->_form->getValidator();
+        if (!$validator->hasField($field)) {
+            return null;
+        }
+        foreach ($validator->field($field)->rules() as $rule) {
+            if ($rule->get('rule') == 'maxLength') {
```
Comment (dereuromark) [inline src/View/Form/FormContext.php]: Same

File: src/View/Helper/FormHelper.php [hunk 10]
```diff
@@ -1476,6 +1476,9 @@ protected function _magicOptions($fieldName, $options, $allowOverride)
             unset($options['step']);
         }
 
+        $validationLength = $context->getMaxLength($fieldName);
+        $fieldDef['length'] = $fieldDef['length'] ?? $validationLength;
```
Comment (dereuromark) [inline src/View/Helper/FormHelper.php]: Why the check here if you already calc based on context?
Cleaner to drop the $validationLength var then and only do the lookup if really necessary.

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 11]
```diff
@@ -1742,24 +1745,24 @@ public function testFormSecurityMultipleControlDisabledFields()
         $result = $this->Form->secure($this->Form->fields);
         $hash = '43c4db25e4162c5e4edd9dea51f5f9d9d92215ec%3AAddresses.0.id%7CAddresses.1.id';
         $tokenDebug = urlencode(json_encode([
-                '/articles/add',
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/FormHelperTest.php]: Please avoid side effects of indentation changes, only minimal diffs, please.
Also, these are technically now incorrect, as the normalized form would have 2 levels here.

File: src/View/Form/EntityContext.php [hunk 12]
```diff
@@ -426,6 +426,29 @@ public function isRequired($field)
...
+    {
+        $parts = explode('.', $field);
+        $validator = $this->_getValidator($parts);
+        $fieldName = array_pop($parts);
+        if (!$validator->hasField($fieldName)) {
+            return null;
+        }
+        foreach ($validator->field($fieldName)->rules() as $rule) {
+            if ($rule->get('rule') === 'maxLength') {
+                return $rule->get('pass')[0] ?? null;
```
Comment (ADmad) [inline src/View/Form/EntityContext.php]: Null coleasing operator `??` is not supported in PHP 5.6.

File: src/View/Form/ArrayContext.php [hunk 13]
```diff
@@ -12,6 +12,7 @@
  * @since         3.0.0
  * @license       https://opensource.org/licenses/mit-license.php MIT License
  */
+
```
Comment (ADmad) [inline src/View/Form/ArrayContext.php]: Please avoid unnecessary whitespace changes.

File: src/View/Form/EntityContext.php (hunk 12 above)
Comment (dereuromark) [reply src/View/Form/EntityContext.php]: Basic isset() ternary suffices.

**Comment by ADmad** [comment]
Since this makes changes to public API it needs to target 3.7 aka the `next` branch.

File: src/View/Form/ArrayContext.php [hunk 14]
```diff
@@ -206,6 +207,31 @@ public function isRequired($field)
...
+            return null;
+        }
+
+        $fieldSchema = Hash::get($this->_context['schema'], $field);
+
+        if(!$fieldSchema){
+            return null;
+        }
+        
+        $maxLength = Hash::get($fieldSchema, 'length');
```
Comment (dereuromark) [inline src/View/Form/ArrayContext.php]: Directly return the result. it is either there or null.

File: src/View/Form/EntityContext.php (hunk 12 above)
Comment (lamasgergo) [reply src/View/Form/EntityContext.php]: > Null coleasing operator `??` is not supported in PHP 5.6.

Would you like support 5.6? I can change it if you would.

**Comment by lamasgergo** [comment]
> Since this makes changes to public API it needs to target 3.7 aka the `next` branch.

I have to add a new pull request to 3.next?

**Comment by dereuromark** [comment]
No, you can switch the target at the top. Whatever you prefer. A clean one to 3.next also wouldnt hurt. Or rebased/squashed, depending on your git foo :)

**Comment by lamasgergo** [comment]
Resubmitted in #12777 .
